Fix crash on PHP 8.3 when a file is missing

XOlegator opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | <!-- comma-separated list of tickets # fixed by the PR, if any -->

This fixes a problem I had after migrating Craft CMS (based on Yii2) to PHP 8.3

XOlegator committed:
Fix crash on PHP 8.3 when a file is missing

samdark commented:
How does the crash looks like? Stacktrace?

XOlegator commented:
```

xicond commented:
Warning
Prior to PHP 8.0.0, the error_reporting() called inside the custom error handler always returned 0 **if the error was suppressed by the @ operator**. As of PHP 8.0.0, it returns the value of this (bitwise) expression: E_ERROR | E_CORE_ERROR | E_COMPILE_ERROR | E_USER_ERROR | E_RECOVERABLE_ERROR | E_PARSE.


As per code [yii\base\ErrorHandler::handleError](https://github.com/yiisoft/yii2/blob/7037fd46abd480330425cab30ce58e180e9d8ead/framework/base/ErrorHandler.php#L248)

```
    public function handleError($code, $message, $file, $line)
    {
        if (error_reporting() & $code) {
            // load ErrorException manually here because autoloading them will not work
            // when error occurs while autoloading a class
            if (!class_exists('yii\\base\\ErrorException', false)) {
                require_once __DIR__ . '/ErrorException.php';
            }
            $exception = new ErrorException($message, $code, $code, $file, $line);

            if (PHP_VERSION_ID < 70400) {
                // prior to PHP 7.4 we can't throw exceptions inside of __toString() - it will result a fatal error
                $trace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS);
                array_shift($trace);
                foreach ($trace as $frame) {
                    if ($frame['function'] === '__toString') {
                        $this->handleException($exception);
                        if (defined('HHVM_VERSION')) {
                            flush();
                        }
                        exit(1);
                    }
                }
            }

            throw $exception;
        }

        return false;
    }
```

So the handleError will show error page at the end for PHP ver >= 8.0.0

rob006 commented:
> As of PHP 8.0.0, it returns the value of this (bitwise) expression: E_ERROR | E_CORE_ERROR | E_COMPILE_ERROR | E_USER_ERROR | E_RECOVERABLE_ERROR | E_PARSE.

`filemtime()` emits `E_WARNING`, so this change doesn't affect this call.

XOlegator commented on framework/caching/FileCache.php:
```diff
@@ -110,7 +110,7 @@ protected function getValue($key)
         $cacheFile = $this->getCacheFile($key);
 
-        if (@filemtime($cacheFile) > time()) {
+        if (file_exists($cacheFile) && @filemtime($cacheFile) > time()) {
```
I looked into the issue a bit deeper. In my Craft CMS setup, I had `DEV_MODE=true` enabled. In this mode, a _PHP Warning_ interrupts program execution just like a _PHP Error_. So, the problem with the missing cache file must be happening at a different level.

Could you clarify what exactly makes the cache file operations **non-atomic** in this case? Yes, there are now two file operations: checking if the file exists and checking its modification time. But isn’t that the whole point of this method? First, verify the file exists, and only if it does, check its modification time. With the double check, we’d simply exit at the first step if the file isn’t there.

rob006 replied:
If you first check if file exist, and then read mtime from it, this means that file could be deleted by other process between these two operations. So `file_exists` doesn't completely protect you from such errors, it only makes them appear much less frequently (at the cost of additional call, which is not necessary in 99.99% setups).

mtangoo replied:
Hi @XOlegator 

> I had DEV_MODE=true enabled. In this mode, a PHP Warning interrupts program execution just like a PHP Error

If this is the case, isn't Yii behaving normal  here? The method  [filemtime](https://www.php.net/manual/en/function.filemtime.php) generates a warning when file does not exist which results in a crash in dev mode. Can you explain where Yii goes wrong?

xicond commented:
> set_error_handler(function($severity, $message, $file, $line) {
>     echo "Error handler called! Message: $message\n";
>     echo "error_reporting() inside handler: " . error_reporting() . "\n"; // 4437 in PHP 8+
>     return true; // Prevent PHP's default error handler from running
> });
> 
> $filename = "non_existent_file.txt";
> $mtime = @filemtime($filename);
> 
> if ($mtime === false) {
>     echo "filemtime failed (returned false)\n";
> }

I'm not sure 

```
Error handler called! Message: filemtime(): stat failed for non_existent_file.txt
error_reporting() inside handler: 4437
filemtime failed (returned false)
```
But I tried the simple test on 8.2, it is shown the error

```
set_error_handler(function($severity, $message, $file, $line) {
    echo "Error handler called! Message: $message\n";
    echo "error_reporting() inside handler: " . error_reporting() . "\n"; // 4437 in PHP 8+
    return true; // Prevent PHP's default error handler from running
});

$filename = "non_existent_file.txt";
$mtime = @filemtime($filename);

if ($mtime === false) {
    echo "filemtime failed (returned false)\n";
}
```

XOlegator committed:
Fix Warning when file is missing

XOlegator committed:
Fix code style - add space

XOlegator commented:
I’ve updated this PR to address the filemtime() warning issue in a more efficient way. The changes:
1.     Replaced @filemtime with set_error_handler to gracefully catch "file not found" warnings without costly file_exists() checks.
2.     Ensured PHP 7.3+ compatibility by using strpos() instead of str_contains().
3.     Fixed both occurrences of this pattern in the codebase.

Why this matters:

-     In Craft CMS’s DEV mode, even warnings (like missing files) trigger exceptions, breaking execution.
-     This solution avoids disk-heavy operations while maintaining silent failure for missing files.
-     No performance penalty vs. @, but more explicit and debuggable.

Let me know if you’d like any adjustments!

rob006 commented:
Can someone explain what specific Craft CMS is doing, that **silenced** warnings are converted to exceptions?

samdark commented:
@brandonkelly would you please point us to the error handing routine of CraftCMS? Thanks.
